orangecontrib/text/widgets/owpreprocess.py

```diff
@@ -1,70 +1,706 @@
+                         QFileDialog, QLineEdit, QCheckBox)
+from nltk.corpus import stopwords
+
+from Orange.widgets import gui, settings
 from Orange.widgets.widget import OWWidget
-from Orange.widgets.settings import Setting
-from Orange.widgets import gui
+from orangecontrib.text.corpus import Corpus
+from orangecontrib.text.preprocess import (PorterStemmer as PS,
+                                           Lemmatizer as LM,
+                                           SnowballStemmer as SS)
 from orangecontrib.text.preprocess import Preprocessor
 
 
+def _i(name, icon_path='icons'):
+    widget_path = os.path.dirname(os.path.abspath(__file__))
+    return os.path.join(widget_path, icon_path, name)
+
+
+class Input:
+    CORPUS = 'Corpus'
+
+
 class Output:
-    PREPROCESSOR = "Preprocessor"
+    PP_CORPUS = 'Corpus'
+
+
+class PreprocessorModule(QWidget):
+    """The base widget for the pre-processing modules."""
+
+    change_signal = Signal()  # Emitted when the settings are changed.
+    # Emitted when the module has a message to display in the main widget.
+    error_signal = Signal(str)
+    enabled = False  # If the module is enabled.
+
+    def __init__(self, title, toggle_enabled, is_enabled):
+        super(PreprocessorModule, self).__init__()
+
+        self.enabled = is_enabled
+
+        # Title bar.
+        title_holder = QWidget()
+        title_holder.setSizePolicy(QSizePolicy.MinimumExpanding,
+                                   QSizePolicy.Fixed)
+        title_holder.setStyleSheet("""
+        .QWidget {
+        background: qlineargradient( x1:0 y1:0, x2:0 y2:1,
+        stop:0 #F8F8F8, stop:1 #C8C8C8);
+        border-bottom: 1px solid #B3B3B3;
+        }
+        """)
+        self.titleArea = QHBoxLayout()
+        self.titleArea.setContentsMargins(15, 10, 15, 10)
+        self.titleArea.setSpacing(0)
+        title_holder.setLayout(self.titleArea)
+
+        self.title_label = QLabel(title)
+        self.title_label.setStyleSheet('font-size: 12px;')
+        self.titleArea.addWidget(self.title_label)
+
+        self.off_label = QLabel('[disabled]')
+        self.off_label.setStyleSheet('color: #B0B0B0; margin-left: 5px;')
+        self.titleArea.addWidget(self.off_label)
+        self.off_label.hide()
+
+        self.titleArea.addStretch()
+
+        if toggle_enabled:
+            self.toggle_module_switch = QCheckBox()
+            self.toggle_module_switch.stateChanged.connect(self.on_toggle)
+            switch_icon_on_resource = _i('on_button.png')
+            switch_icon_off_resource = _i('off_button.png')
+            style_sheet = '''
+            QCheckBox::indicator {
+                width: 23px;
+                height: 23px;
+            }
+            QCheckBox::indicator:checked {
+                image: url(%s);
+            }
+            QCheckBox::indicator:unchecked {
+                image: url(%s);
+            }
+            ''' % (switch_icon_off_resource, switch_icon_on_resource)
```

Comment:
This works fine, but seem counterintuitive. How is `switch_icon_off_resource ` an URL for `QCheckBox::indicator:checked`? Shouldn't it be an URL for `unchecked `?